File: 3-1. Algorithm Analysis and Design/Lab/Lab 7/7-3.cpp
#include <iostream>
#include <string>
using namespace std;

// Node structure for the Huffman Tree
struct Node {
    char data;         // Character stored in the node (for leaf nodes)
    int freq;          // Frequency of the character
    Node* left;        // Left child
    Node* right;       // Right child

    Node(char data, int freq, Node* left = nullptr, Node* right = nullptr)
        : data(data), freq(freq), left(left), right(right) {}
};

// Function to decode the Huffman encoded string
string decode_huff(Node* root, const string& s) {
    string decoded_string = "";
    Node* current = root;

    for (char bit : s) {
        if (bit == '0') {
            current = current->left;  // Move left for '0'
        } else {
            current = current->right;  // Move right for '1'
        }

        // Check if current node is a leaf node
        if (!current->left && !current->right) {
            decoded_string += current->data;  // Append the character
            current = root;  // Reset to root for the next character
        }
    }

    return decoded_string;
}

// Example usage
int main() {
    // Construct the sample Huffman tree
    Node* root = new Node('\0', 12);  // Root node with null data
    root->left = new Node('B', 5);  // Left child
    root->right = new Node('\0', 7);  // Internal node
    root->right->left = new Node('A', 3);  // Left child of right node
    root->right->right = new Node('C', 4);  // Right child of right node

    // Encoded string
    string encoded_string = "1001011";

    // Decode the string
    string decoded_result = decode_huff(root, encoded_string);

    // Output the decoded string
    cout << "Decoded string: " << decoded_result << endl;

    // Free allocated memory (for the sake of good practice)
    delete root->left;
    delete root->right->left;
    delete root->right->right;
    delete root->right;
    delete root;

    return 0;
}

File: 3-1. Algorithm Analysis and Design/Lab/Lab 7/7-3.py
class Node:
    def __init__(self, char=None, freq=None):
        self.char = char  # Character for this node
        self.freq = freq  # Frequency of the character
        self.left = None  # Left child
        self.right = None  # Right child

def decode_huff(root, s):
    decoded_string = []
    current = root  # Start from the root of the tree
    
    for bit in s:
        if bit == '0':
            current = current.left  # Move to the left child
        else:
            current = current.right  # Move to the right child
        
        # If we reach a leaf node, add the character to the decoded string
        if current.left is None and current.right is None:
            decoded_string.append(current.char)
            current = root  # Reset to the root for the next character
            
    return ''.join(decoded_string)

# Example usage:
# Build the Huffman tree for the given problem:
root = Node()
root.left = Node('B')
root.right = Node('A')
root.right.left = Node('R')
root.right.right = Node()
root.right.right.left = Node('C')
root.right.right.right = Node('D')

# Decoded input string
s = "1001011"
decoded_result = decode_huff(root, s)
print(decoded_result)  # Output: ABACA
